Improve reduce pattern performance with mid-sized inputs

julianmi opened the pull request:
This patch refactors the reduce pattern by:
- Fixed numbers of `__iters_per_work_item` are used independent of the device. This unifies the similar `transform_reduce_known` and `transform_reduce_unknown` structs.
- The sequential kernel was removed since it does not provide significant performance improvements.
- The performance of mid-sized arrays is improved by reducing the `__iters_per_work_item` for the second kernel to the lowest value needed. This better balances the work across work items. If each work item has less than one element to work on, the work group size is further reduced if possible. This lowers the complexity of the work group reduction.
- Improved documentation and naming.

mmichel11 commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h:
```diff
@@ -113,30 +139,74 @@ struct __parallel_transform_reduce_small_submitter<__work_group_size, __iters_pe
+
+            __cgh.parallel_for<_LeafName...>(
+                sycl::nd_range<1>(sycl::range<1>(__work_group_size2), sycl::range<1>(__work_group_size2)),
+                [=](sycl::nd_item<1> __item_id) {
```
This kernel is identical to the kernel used in `__parallel_transform_reduce_small_submitter`. Would it make sense to unify this lambda body in a separate functor so that this code is not duplicated?

julianmi replied:
Thanks, I will change that.

mmichel11 commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h:
```diff
@@ -298,114 +364,93 @@ __parallel_transform_reduce(_ExecutionPolicy&& __exec, _ReduceOp __reduce_op, _T
 
-    if (__n <= 65536 && __work_group_size >= 512)
+    // Use single work group implementation if array < __work_group_size * __iters_per_work_item.
+    if (__n <= 256 && __work_group_size >= 256)
```
Would it be a better practice to turn all subsequent `if` statements into `else if` statements?

julianmi replied:
I don't have a strong opinion on this but `else if` might be easier to read.

danhoeflinger replied:
Doesn't `else if` avoid a lot of unnecessary branching for a single control flow (once you hit your `true` case, you skip all the rest of the branching)?
Unless we have a explicit need to have consecutive `if` statements, I think `else if` is correct.

julianmi replied:
While I agree in general, this routine uses return statements in the if statements.

danhoeflinger replied:
Ah yes, my mistake.  I'll still put my vote in for `else if` because that is the norm here I think.
However, I'll retract my previous statement.

julianmi replied:
I switched to `else if` statements.

danhoeflinger replied:
Also for this set of `if`/`else if` statements:  They are all checking `__work_group_size >=256`, does it make sense to pull this out and above?

mmichel11 commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h:
```diff
@@ -113,30 +139,74 @@ struct __parallel_transform_reduce_small_submitter<__work_group_size, __iters_pe
+            if (__n < __work_group_size)
+            {
+                __work_group_size2 = __n;
+                if ((__work_group_size2 & (__work_group_size2 - 1)) != 0)
```
I see in our codebase that we have `__dpl_bit_floor` but no equivalent `__dpl_bit_ceil`. Am I correct that this if statement, bit floor, and bit shift could be replaced with a single bit ceil operation?

Since `std::bit_ceil` is in the C++20 standard (https://en.cppreference.com/w/cpp/numeric/bit_ceil), it may be useful to add an equivalent `__dpl_bit_ceil` into our utilities.

julianmi replied:
This is correct. I will change that to `__dpl_bit_ceil ` once #909 is merged.

danhoeflinger commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h:
```diff
@@ -113,102 +170,124 @@ struct __parallel_transform_reduce_small_submitter<__work_group_size, __iters_pe
         auto __reduce_pattern = unseq_backend::reduce_over_group<_ExecutionPolicy, _ReduceOp, _Tp>{__reduce_op};
 
-        const _Size __n_items = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __iters_per_work_item);
+        _Size __size_per_work_group =
```
Can this be `constexpr`?

julianmi replied:
Done.

danhoeflinger commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h:
```diff
@@ -113,102 +170,124 @@ struct __parallel_transform_reduce_small_submitter<__work_group_size, __iters_pe
+            auto __res_acc = __res.template get_access<sycl::access_mode::write>(__cgh);
+            __dpl_sycl::__local_accessor<_Tp> __temp_local(sycl::range<1>(__work_group_size2), __cgh);
+
+            __cgh.parallel_for<_LeafName...>(
```
We are using `_LeafName...` as the name for this kernel which includes a `work_group_size` which may deviate from the actually used one:  `work_group_size2` which can depend on `__n`.  Can this get us into trouble?  
If not, should `work_group_size` be removed from the name?

julianmi replied:
Good point. I think we should remove `work_group_size` from the generated leaf kernel name since it has no meaning. I think we can even remove `work_group_size` from all kernel names since we only use 256.

julianmi commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h:
```diff
@@ -113,102 +170,124 @@ struct __parallel_transform_reduce_small_submitter<__work_group_size, __iters_pe
+        _Size __n_items = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __iters_per_work_item1);
 
-        sycl::buffer<_Tp> __res(sycl::range<1>(1));
+        sycl::buffer<_Tp> __temp((sycl::range<1>(__n_groups)));
```
This solves the ambiguity of the targeted variable definition vs. a potential function declaration.

danhoeflinger replied:
Ah yes, thanks for clarification.

julianmi replied:
I switched to curly braces for clarity.

mmichel11 commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h:
```diff
@@ -34,25 +34,70 @@ namespace __par_backend_hetero
 
-template <typename _Tp, typename... _Name>
-struct __parallel_transform_reduce_seq_submitter<_Tp, __internal::__optional_kernel_name<_Name...>>
+template <::std::uint16_t __work_group_size, ::std::size_t __iters_per_work_item, typename _Tp, typename... _Name>
```
Since the input size is small here, can `__iters_per_work_item` be a type smaller than `std::size_t`?

dmitriy-sobolev commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h:
```diff
@@ -288,7 +358,15 @@ struct __parallel_transform_reduce_impl
 }; // struct __parallel_transform_reduce_impl
 
-// General version of parallel_transform_reduce. Calls optimized kernels.
+// General version of parallel_transform_reduce. Processes elements in order to enable non-commutative algorithms.
```
I do not quite understand the meaning of "Processes elements in order to enable non-commutative algorithms". Does it mean that this implementation provides a deterministic result when `_ReduceOp` predicate is non-commutative?

It may use `sycl::reduce_over_group` which has the following note:
> The result of a call to these functions is non-deterministic if the binary operator is not commutative and associative

julianmi replied:
Exactly. This version does not require commutative operators. I tried to clarify the comment.

`sycl::reduce_over_group` is used for the group reduction if the type is arithmetic and the operator is commutative (cf. `__has_known_identity`). We process elements in order otherwise.

dmitriy-sobolev replied:
Thanks for the clarification! 

The following is not related to the PR and thus does not block it. I was not able to find a statement that the predicates with available known identities are both associative and commutative. `sycl::minimum`, one of these predicates, returns
the first argument when the arguments are equivalent, so the result for `sycl::minimum(a, b)` and `sycl::minimum(b, a)` are different when `a == b`. Could you open an issue if you think it is a valid case?

julianmi replied:
I agree with @danhoeflinger. The spec for `sycl::reduce_over_group` (https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_reduce) specifies the binary operators in https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:function-objects which include `sycl::minimum` and `sycl::maximum`. The only problem I see is with algorithms that use minimum or maximum but the order matters such as for `min_element()` where the first index with that value needs to be returned. Our implementation uses a user-defined type then and we fall back to an in-order group reduction.

danhoeflinger replied:
Let me make a minor correction to what I said:  for `sycl::minimum` and `sycl::maximum` it `has_known_identity` specifically for integral types, floating point types and sycl::half rather than all arithmetic types.  This leaves out complex types (which is good).

dmitriy-sobolev commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h:
```diff
@@ -34,25 +34,70 @@ namespace __par_backend_hetero
+void
+__device_reduce_kernel(const _NDItemId __item_id, const _Size __n, const _Size __n_items,
+                       const _TransformPattern __transform_pattern, const _ReducePattern __reduce_pattern,
+                       const _InitType __init, const _AccLocal& __local_mem, _Tmp& __temp_acc, const _Acc&... __acc)
```
Const is missing for `_Tmp& __temp_acc`, `__init` is not used.

julianmi replied:
Applied.

julianmi commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h:
```diff
@@ -61,50 +106,61 @@ struct __parallel_transform_reduce_seq_submitter<_Tp, __internal::__optional_ker
-template <::std::uint16_t __work_group_size, ::std::size_t __iters_per_work_item, typename _Tp, typename... _Name>
-struct __parallel_transform_reduce_small_submitter<__work_group_size, __iters_per_work_item, _Tp,
-                                                   __internal::__optional_kernel_name<_Name...>>
+// Parallel_transform_reduce for a mid-sized arrays using two reduction steps.
```
Fixed.

julianmi commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h:
```diff
@@ -61,50 +106,61 @@ struct __parallel_transform_reduce_seq_submitter<_Tp, __internal::__optional_ker
-            __reduce_op, _TransformOp{__transform_op}};
+        auto __transform_pattern =
+            unseq_backend::transform_reduce<_ExecutionPolicy, __iters_per_work_item, _ReduceOp, _TransformOp>{
+                __reduce_op, _TransformOp{__transform_op}};
```
Applied.

julianmi commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h:
```diff
@@ -113,102 +169,122 @@ struct __parallel_transform_reduce_small_submitter<__work_group_size, __iters_pe
         sycl::event __reduce_event = __exec.queue().submit([&, __n, __n_items](sycl::handler& __cgh) {
             oneapi::dpl::__ranges::__require_access(__cgh, __rngs...); // get an access to data under SYCL buffer
-            auto __res_acc = __res.template get_access<access_mode::write>(__cgh);
+            auto __temp_acc = __temp.template get_access<sycl::access_mode::write>(__cgh);
```
I did not in my tests, but I changed it to provided additional information to the compiler. It might be useful on different platforms or in the future.

julianmi commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h:
```diff
@@ -113,102 +169,122 @@ struct __parallel_transform_reduce_small_submitter<__work_group_size, __iters_pe
+        }
+        __n_items = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __iters_per_work_item2);
+
+        sycl::buffer<_Tp> __res(sycl::range<1>(1));
```
Allocating a buffer of size `__tmp` + 1 for the result is faster than allocation two separate buffers. However, the current implementation transfers whole buffers instead of sub-buffers. The additional transfer costs are higher than the allocation costs when using an extended `__tmp` buffer.

julianmi commented:
I have addressed the comments so far except for using `__dpl_bit_ceil ` which requires #909 to be merged.

I have further verified that this patch improves the performance on CPU and GPU devices. Thus, the previous CPU-specific code path with distributing one work group per compute unit is not required anymore. This patch will use more work groups which might even be beneficial to hide latencies.

The clang-format hints are unrelated to the code changed in this PR.

dmitriy-sobolev commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h:
```diff
@@ -34,181 +34,258 @@ namespace __par_backend_hetero
+            {
+                __work_group_size2 = __n;
+                if ((__work_group_size2 & (__work_group_size2 - 1)) != 0)
+                    __work_group_size2 = oneapi::dpl::__internal::__dpl_bit_floor(__work_group_size2) << 1;
```
`__work_group_size2` may become larger than `__work_group_size`. It may result in allocation more SLM than available on the device.

julianmi replied:
I don't think it will. `__work_group_size` is set to a power of two (currently 256). `__work_group_size2` uses the bit ceiling of `__n` only if `__n` is smaller than `__work_group_size`. The maximum of `__work_group_size2` should thus be `__work_group_size`.

The goal of this code is to reduce the complexity of the work group reduction if a smaller number of preliminary results from the first kernel than the `__work_group_size` need to be reduced in the second kernel.

dmitriy-sobolev reviewed (approved):
LGTM. 

I think, a better title for the PR is "Improve reduce pattern performance with mid-sized inputs".

rarutyun commented on include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h:
```diff
@@ -34,181 +34,258 @@ namespace __par_backend_hetero
-__parallel_transform_reduce_small_impl(_ExecutionPolicy&& __exec, _Size __n, _ReduceOp __reduce_op,
-                                       _TransformOp __transform_op, _InitType __init, _Ranges&&... __rngs)
+__parallel_transform_reduce_mid_impl(_ExecutionPolicy&& __exec, _Size __n, const _ReduceOp& __reduce_op,
+                                     const _TransformOp& __transform_op, const _InitType& __init, _Ranges&&... __rngs)
```
Just curious why Callables are now passed by `const` reference. Why is it better (keeping in mind the in the APIs all user callables come by value)?
